auth/src/main/java/com/firebase/ui/auth/ui/idp/CredentialSignInHandler.java

```diff
@@ -82,37 +93,52 @@ public void onComplete(@NonNull Task<AuthResult> task) {
                             .addOnFailureListener(new OnFailureListener() {
                                 @Override
                                 public void onFailure(@NonNull Exception e) {
-                                    // TODO: What to do when signing in with Credential fails
-                                    // and we can't continue to Welcome back flow without
-                                    // knowing providers?
+                                    mHelper.finishActivity(
+                                            mActivity,
+                                            ResultCodes.CANCELED,
+                                            IdpResponse.getErrorCodeIntent(ErrorCodes.UNKNOWN_ERROR));
                                 }
                             });
                     return;
                 }
-            } else {
-                Log.e(TAG, "Unexpected exception when signing in with credential",
-                      task.getException());
             }
+            Log.e(TAG, "Unexpected exception when signing in with credential", task.getException());
             mHelper.dismissDialog();
         }
     }
 
     private class StartWelcomeBackFlow implements OnSuccessListener<ProviderQueryResult> {
         @Override
         public void onSuccess(@NonNull ProviderQueryResult result) {
+            AuthCredential credential = AuthCredentialHelper.getAuthCredential(mResponse);
+            if (mHelper.canLinkAccounts()
+                    && credential != null
+                    && result.getProviders().contains(credential.getProvider())) {
+                // We don't want to show the welcome back dialog since the user selected
+                // an existing account and we can just link the two accounts without knowing
+                // prevCredential.
+                AccountLinker.link(
+                        mActivity,
+                        mHelper,
+                        mResponse,
+                        credential,
+                        null);
+                return;
+            }
+
             mHelper.dismissDialog();
 
-            String provider = result.getProviders().get(0);
+            String provider = getHighestPriorityProvider(result.getProviders());
```

Comment:
`get(0)` was returning Twitter even though google was available which is completely unacceptable.